### PR TITLE
fixing advanced indexing operation for empty arrays

### PR DESCRIPTION
**Pull request by ipdemes**
This PR fixes test cases when input/index and/or values are empty arrays 
It depends on https://github.com/nv-legate/legate.core/pull/319 to me merged first

### CODE REVIEW

**Comment by ipdemes** [comment]
This PR also fixes `reshape` for empty arrays with  ndim >1 (arrays with shapes (1,0), (1,0,0) e.t.c)

File: cunumeric/deferred.py [hunk 1]
```diff
@@ -503,6 +503,25 @@ def _create_indexing_array(
                         f"dimension {i} doesn't match to the shape of the"
                         f"index array which is {rhs.shape[i]}"
                     )
+
+            if key.size == 0:
+                rhs_store = rhs.base
+                for i in range(key.ndim - 1):
+                    rhs_store = rhs_store.project(0, 0)
```
Comment (manopapad) [inline cunumeric/deferred.py:510]: This will fail if one of the first `key.ndim - 2` dimensions on `rhs` is zero-size.

File: cunumeric/deferred.py [hunk 2]
```diff
@@ -503,6 +503,25 @@ def _create_indexing_array(
                         f"dimension {i} doesn't match to the shape of the"
                         f"index array which is {rhs.shape[i]}"
                     )
+
+            if key.size == 0:
```
Comment (manopapad) [inline cunumeric/deferred.py:507]: This code is never exercised in the current test suite, because it is quite hard to produce 0-size `DeferredArray`s (cunumeric prefers 0-size arrays to be Eager). I wrote some code that thoroughly exercises this part, and it seems to be failing currently:

```
import numpy as np
import cunumeric as cn

from legate.core import LEGATE_MAX_DIM

def mk_deferred_array(lib, shape):
    if np.prod(shape) != 0:
        return lib.ones(shape)
    # for shape (2,0,3,4): good_shape = (2,1,3,4)
    good_shape = tuple(max(1, dim) for dim in shape)
    # for shape (2,0,3,4): key = [:,[False],:,:]
    key = tuple([False] if dim == 0 else slice(None) for dim in shape)
    return lib.ones(good_shape)[key]

for arr_ndim in range(1, LEGATE_MAX_DIM + 1):
    for idx_ndim in range(1, arr_ndim + 1):
        for zero_dim in range(arr_ndim):
            arr_shape = tuple(0 if dim == zero_dim else 3 for dim in range(arr_ndim))
            np_arr = mk_deferred_array(np, arr_shape)
            cn_arr = mk_deferred_array(cn, arr_shape)
            idx_shape = arr_shape[:idx_ndim]
            val_shape = arr_shape if idx_ndim == 1 else (np.prod(idx_shape),) + arr_shape[idx_ndim:]
            np_idx = np.ones(idx_shape, dtype=np.bool_)
            cn_idx = cn.ones(idx_shape, dtype=np.bool_)
            assert np.array_equal(np_arr[np_idx], cn_arr[cn_idx])
            np_val = np.random.random(val_shape)
            cn_val = cn.array(np_val)
            np_arr[np_idx] = np_val
            cn_arr[cn_idx] = cn_val
            assert np.array_equal(np_arr, cn_arr)
```

It currently fails with:

```
[0 - 70000b0c4000]    1.421775 {5}{runtime}: [error 209] LEGION ERROR: The destination indirect field for a copy operation has the incorrect size for the destination region coordinate space. Field 1048595 of destination indirect region requirement 0 is 8 bytes but the coordinate types of the destination space is 16 bytes for copy operation (UID 62) in parent task legion_python_main (UID 1). (from file /Users/mpapadakis/legate.core/legion/runtime/legion/legion_ops.cc:4648)
```

Please make sure it passes, and add it to the test suite.

File: cunumeric/deferred.py [hunk 3]
```diff
@@ -503,6 +511,32 @@ def _create_indexing_array(
...
+                out = cast(
+                    DeferredArray,
+                    self.runtime.create_empty_thunk(
+                        out_shape,
+                        rhs.dtype,
+                        inputs=[rhs],
+                    ),
+                )
+
+                out._fill_with_zero()
```
Comment (ipdemes) [inline cunumeric/deferred.py:537]: since out is uninitialized, we need to `touch it` with write permissions even if its size is zero. I use `FILL` task for this

File: cunumeric/deferred.py [hunk 4]
```diff
@@ -631,8 +666,12 @@ def _create_indexing_array(
                     # in case of the mixed indises we all nonzero
                     # for the bool array
                     k = k.nonzero()
-                    shift += len(k) - 1
-                    tuple_of_arrays += k
+                    if k[0].size == 0:
+                        store = store.project(dim + shift, 0)
+                        store = store.promote(dim + shift, 0)
```
Comment (ipdemes) [inline cunumeric/deferred.py:671]: if there is no `true` elements in the boolean array, we will just transform the original array

File: cunumeric/deferred.py [hunk 5]
```diff
@@ -830,13 +870,15 @@ def set_item(self, key: Any, rhs: Any) -> None:
             if lhs.base.kind == Future:
                 lhs = self._convert_future_to_store(lhs)
 
-            copy = self.context.create_copy()
-            copy.set_target_indirect_out_of_range(False)
+            if index_array.size != 0:
```
Comment (ipdemes) [inline cunumeric/deferred.py:874]: We don't need to perform copy is indirect array is zero-size

File: src/cunumeric/index/advanced_indexing.cc [hunk 6]
```diff
@@ -79,11 +79,6 @@ struct AdvancedIndexingImplBody<VariantKind::CPU, CODE, DIM, OUT_TYPE> {
       if (index[p] == true) { size++; }
     }
 
-    if (0 == size) {
```
Comment (ipdemes) [inline src/cunumeric/index/advanced_indexing.cc:82]: I remove this block since we our output region should be specific shape even if its size is 0

File: cunumeric/deferred.py (hunk 3 above)
Comment (manopapad) [reply cunumeric/deferred.py:537]: Is it sufficient to do something like this?

```
out.fill(np.zeros((), dtype=out.dtype))
```

Then we don't need to add an extra argument to the FILL task.

Comment (ipdemes) [reply cunumeric/deferred.py:537]: Yes, you are right. This is so simple! I have spent quite some time thinking how to best implement this and missed the fact that we can pass numpy arrays to deferred methods

File: cunumeric/deferred.py (hunk 4 above)
Comment (manopapad) [reply cunumeric/deferred.py:662]: I see the optimization opportunity here, but note that, since `nonzero()` has an unknown output size, the check on `k[0].size` has to block until `nonzero()` returns, even if the call happens to return a non-empty array, and we would rather avoid blocking in the top-level program.

I think the original code here would still work, we would just get an empty indices array out of `_zip_indices`, does that sound right?

File: cunumeric/deferred.py [hunk 7]
```diff
@@ -1230,6 +1263,7 @@ def _fill(self, value: Any) -> None:
             task.add_output(self.base)
             task.add_input(value)
             task.add_scalar_arg(argval, bool)
+            task.add_scalar_arg(False, bool)  # fill with zero
```
Comment (manopapad) [inline cunumeric/deferred.py:1266]: I think this is not necessary anymore.

File: tests/integration/test_advanced_indexing.py [hunk 8]
```diff
@@ -81,6 +95,98 @@ def test_setitem_scalar_1d(arr, idx, val):
...
+
+
+def mk_deferred_array(lib, shape):
+    if np.prod(shape) != 0:
+        return lib.ones(shape)
+    # for shape (2,0,3,4): good_shape = (2,1,3,4)
+    good_shape = tuple(max(1, dim) for dim in shape)
+    # for shape (2,0,3,4): key = [:,[False],:,:]
+    key = tuple([False] if dim == 0 else slice(None) for dim in shape)
+    print("IRINA DEBUG good_shape , key", good_shape, key)
```
Comment (manopapad) [inline tests/integration/test_advanced_indexing.py:119]: Probably want to remove these debug statements (also below), to avoid polluting the test output.

File: tests/integration/test_advanced_indexing.py [hunk 9]
```diff
@@ -81,6 +95,98 @@ def test_setitem_scalar_1d(arr, idx, val):
...
+    return lib.ones(good_shape)[key]
+
+
+def test_zero_size():
```
Comment (manopapad) [inline tests/integration/test_advanced_indexing.py:123]: @bryevdv Is there a way to write this with fixtures, where the fixture for `arr_ndim` ranges independently from 1 to `LEGATE_MAX_DIM + 1`, and the fixtures for `idx_ndim` and `zero_dim` depend on the chosen value of `arr_ndim`?

Comment (bryevdv) [reply tests/integration/test_advanced_indexing.py:123]: `pytest.mark.parametrize` can accept a callable that returns an iterable. So I'd suggest writing a little helper function that yields the desired argument tuples.

File: cunumeric/array.py [hunk 10]
```diff
@@ -3019,6 +3019,17 @@ def ravel(self, order="C") -> ndarray:
...
+            for i in range(self.ndim):
+                if self.shape[i] != 0:
+                    idx += (0,)
+                else:
+                    idx += (slice(None),)
+            idx = tuple(idx)
+            self = self[idx].copy()
+            return self
+
         return self.reshape(-1, order=order)
```
Comment (manopapad) [inline cunumeric/array.py:3033]: I think just falling back to `self.reshape(-1, order=order)` works, and the special-case isn't necessary.

File: cunumeric/deferred.py (hunk 4 above)
Comment (ipdemes) [reply cunumeric/deferred.py:662]: Yes, original code will work. Removing this special case

File: cunumeric/deferred.py (hunk 7 above)
Comment (ipdemes) [reply cunumeric/deferred.py:1266]: right

File: cunumeric/array.py (hunk 10 above)
Comment (ipdemes) [reply cunumeric/array.py]: I have added the test case that wouldn't work with the original implementation and moved this special case to `reshape`

File: cunumeric/deferred.py [hunk 11]
```diff
@@ -651,6 +677,8 @@ def _create_indexing_array(
             # the store with transformation
             rhs = cast(DeferredArray, self._copy_store(store))
 
+        if len(tuple_of_arrays) == 0:
+            return False, rhs, rhs, self
```
Comment (manopapad) [inline cunumeric/deferred.py:681]: I don't think this can ever happen anymore, since the special case was removed above. We could instead just
```
assert(len(len_tuple_of_arrays) > 0)
```
to make this clear on the code.

File: tests/integration/test_advanced_indexing.py (hunk 9 above)
Comment (ipdemes) [reply tests/integration/test_advanced_indexing.py]: thank you @bryevdv ! I have updated the test accordingly

File: tests/integration/test_advanced_indexing.py [hunk 12]
```diff
@@ -81,6 +95,95 @@ def test_setitem_scalar_1d(arr, idx, val):
...
+    return lib.ones(good_shape)[key]
+
+
+def gen_args():
+    result = []
+    for arr_ndim in range(1, LEGATE_MAX_DIM + 1):
+        for idx_ndim in range(1, arr_ndim + 1):
+            for zero_dim in range(arr_ndim):
+                result += [[arr_ndim, idx_ndim, zero_dim]]
+    return result
```
Comment (bryevdv) [inline tests/integration/test_advanced_indexing.py:128]: FYI this should also work:
```suggestion
def gen_args():
    for arr_ndim in range(1, LEGATE_MAX_DIM + 1):
        for idx_ndim in range(1, arr_ndim + 1):
            for zero_dim in range(arr_ndim):
                yield arr_ndim, idx_ndim, zero_dim
```

File: cunumeric/deferred.py (hunk 11 above)
Comment (ipdemes) [reply cunumeric/deferred.py]: probably not. removing this case

File: tests/integration/test_advanced_indexing.py (hunk 12 above)
Comment (ipdemes) [reply tests/integration/test_advanced_indexing.py]: Thank you!

**Review by manopapad — APPROVED** [review]
Great job fixing this very stupid corner case

**Comment by ipdemes** [comment]
Thank you, Manolis! Will merge it after CI passes
